File: algorithms/code/leetcode/lc349_intersection_of_two_arrays/lc349_intersection_of_two_arrays.py
#!/usr/bin/env python3

# https://leetcode.com/problems/intersection-of-two-arrays/

import unittest

from typing import List


class Solution:
    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        nums1.sort()
        nums2.sort()
        found = set()
        i1 = 0
        i2 = 0
        while i1 < len(nums1) and i2 < len(nums2):
            if nums1[i1] < nums2[i2]:
                i1 += 1
                continue
            if nums1[i1] > nums2[i2]:
                i2 += 1
                continue
            found.add(nums1[i1])
            i1 += 1
            i2 += 1
        return list(found)


class TestCode(unittest.TestCase):
    def test_example(self):
        nums1 = [1, 2, 2, 1]
        nums2 = [2, 2]
        result = Solution().intersection(nums1, nums2)
        result.sort()
        expected = [2]
        self.assertListEqual(expected, result)

    def test_example_flipped(self):
        nums1 = [2, 2]
        nums2 = [1, 2, 2, 1]
        result = Solution().intersection(nums1, nums2)
        result.sort()
        expected = [2]
        self.assertListEqual(expected, result)

    def test_12_11(self):
        expected = [1]
        result = Solution().intersection([1, 2], [1, 1])
        result.sort()
        self.assertListEqual(expected, result)

    def test_495_94985(self):
        expected = [4, 5, 9]
        result = Solution().intersection([4, 9, 5], [9, 4, 9, 8, 5])
        result.sort()
        self.assertListEqual(expected, result)

File: algorithms/java/src/test/java/leetcode/LC349IntersectionOfTwoArrays.java
package leetcode.lc349_intersection_of_two_arrays;

import org.junit.jupiter.api.Test;


import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;

/**
 * https://leetcode.com/problems/intersection-of-two-arrays/
 * #easy
 */
public final class LC349IntersectionOfTwoArrays {
    public final class Solution {
        public int[] intersection(final int[] nums1, final int[] nums2) {
            Arrays.sort(nums1);
            Arrays.sort(nums2);
            Set<Integer> found = new HashSet<>();
            for (int i1 = 0, i2 = 0; i1 < nums1.length && i2 < nums2.length; ) {
                if (nums1[i1] < nums2[i2]) {
                    i1++;
                    continue;
                }
                if (nums1[i1] > nums2[i2]) {
                    i2++;
                    continue;
                }
                found.add(nums1[i1]);
                i1++;
                i2++;
            }
            int[] result = new int[found.size()];
            int index = 0;
            for (int value : found) {
                result[index] = value;
                index++;
            }
            return result;
        }
    }

    @Test
    public void test_example() throws Exception {
        int[] nums1 = {1, 2, 2, 1};
        int[] nums2 = {2, 2};
        int[] result = new Solution().intersection(nums1, nums2);
        Arrays.sort(result);
        int[] expected = {2};
        assertArrayEquals(expected, result);
    }
}
